config/samples/cni_v1alpha1_calicoconfig.yaml

```diff
@@ -0,0 +1,10 @@
+apiVersion: cni.tanzu.vmware.com/v1alpha1
+kind: CalicoConfig
+metadata:
+  name: calicoconfig-sample
+spec:
+  Namespace: "kube-system"
+  InfraProvider: vsphere
+  IpFamily: ipv4
+  ClusterCIDR: "192.168.0.0/16,fd00:100:96::/48"
+  VethMTU: 0
```

Comment:
@vijaykatam Do you think it matters if the spec be like this (without calico.config nesting) as opposed to the following as in the values.yaml in the[ upstream TCE](https://github.com/vmware-tanzu/community-edition/blob/main/addons/packages/calico/3.19.1/bundle/config/values.yaml)?

```
spec:
  Namespace: "kube-system"
  InfraProvider: vsphere
  IpFamily: ipv4
  calico:
    config:
       ClusterCIDR: "192.168.0.0/16,fd00:100:96::/48"
       VethMTU: 0
```

Reply:
Personally, I think the format should match in both these 2 places, Otherwise, it will introduce trouble when doing conversion. We either change the schema in TCE or update the formats in here.

Reply:
We should not expose clusterCIDR as a parameter as it will be derived from [ClusterNetwork](https://github.com/kubernetes-sigs/cluster-api/blob/2f975793bf79f8e40173eecd0999cab2d8f17b9d/api/v1beta1/cluster_types.go#L182).

Reply:
I think we need to have some conversion logic that is shared and used by all the config controllers. This logic will convert fields one by one correspondingly from CRD to the yaml used in data values secret.
On top of that, we add special logic to each controller to handle things like `ClusterNetwork` that Vijay mentioned.

Reply:
@blc1996 , Is this logic for converting fields from CRD to yaml in data values secret the same thing what @vijaykatam was mentioning, that it should be automatically generated from CRD?

Reply:
Firstly, if we want to use the [yam.marshall/yaml.unmarshall](https://pkg.go.dev/gopkg.in/yaml.v2#Unmarshal) functions provided by GO to generates the content we put into data values secret

```
Namespace: "kube-system"
InfraProvider: vsphere
IpFamily: ipv4
calico:
    config:
	ClusterCIDR: "192.168.0.0/16,fd00:100:96::/48"
	VethMTU: 0
```

We would need to define some struct similar to the following

```
type CalicoConfigYaml struct {

	Namespace string `yaml:"namespace,omitempty"`

	InfraProvider string `yaml:"infraProvider"`

	IPFamily string `yaml:"ipFamily,omitempty"`

	Config struct `yaml:"vethMTU,omitempty"` {
		VethMTU      int64 `yaml:"vethMTU,omitempty"`
                ClusterCIDR string `yaml:"clusterCIDR,omitempty"`
	}
}
```

It is pretty similar to the struct defined for Kubebuilder

```
// CalicoConfigSpec defines the desired state of CalicoConfig
type CalicoConfigSpec struct {

	// The namespace in which calico is deployed
	// +kubebuilder:validation:Optional
	// +kubebuilder:default:=kube-system
	Namespace string `json:"namespace,omitempty"`

	// Infrastructure provider in use
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:Enum="aws";"azure";"vsphere";"docker"
	InfraProvider string `json:"infraProvider"`

	// The IP family calico should be configured with
	// +kubebuilder:validation:Optional
	// +kubebuilder:validation:Enum="ipv4";"ipv6";"ipv4,ipv6";"ipv6,ipv4"
	// +kubebuilder:default:=ipv4
	IPFamily string `json:"ipFamily,omitempty"`

	// Maximum transmission unit setting
	// +kubebuilder:validation:Optional
	// +kubebuilder:validation:Minimum=0
	// +kubebuilder:default:=0
	VethMTU int64 `json:"vethMTU,omitempty"`
}
```

We would need to do a mapping of values between `CalicoConfigYaml` and `CalicoConfigSpec`. If we keep them as similar as possible, it would be easier for the conversion.

Reply:
Changed to use nested structs fo ```VethMTU``` to align with [upstream TCE](https://github.com/vmware-tanzu/community-edition/blob/main/addons/packages/calico/3.19.1/bundle/config/values.yaml) and easier conversion as suggested above